scrapy_poet/page_input_providers.py

```diff
@@ -361,13 +390,24 @@ async def __call__(
             )
 
             try:
-                po_instances = await self.injector.build_instances(
-                    request, response, plan
+                deferred_or_future = maybe_deferred_to_future(
+                    self.injector.build_instances(request, response, plan)
                 )
+                # RecursionError NOT raised when ``AsyncioSelectorReactor`` is used.
+                # Could be related: https://github.com/python/cpython/issues/93837
+
+                # Need to check before awaiting on the ``asyncio.Future``
+                # before it gets stuck on a potential deadlock.
+                if asyncio.isfuture(deferred_or_future):
+                    if self.check_if_deadlock(request):
+                        raise ProviderDependencyDeadlockError(
+                            self.template_deadlock_msg.format(plan=plan)
+                        )
+
+                po_instances = await deferred_or_future
             except RecursionError:
```

Comment:
I'm thinking about removing this try-except block which checks for `RecursionError` and simply use `check_if_deadlock()` for both Futures and Deferreds. However, I also think that there's a much better workaround for `ayncio` that I haven't found yet. 

Might be worth having 2 ways to catch deadlocks here for Futures and Deferreds for the time being.